Sources/ISO8601DurationFormatter/ISO8601DurationFormatter.swift

```diff
@@ -23,55 +38,77 @@ public class ISO8601DurationFormatter: Formatter {
     - parameter string: A [String](https://developer.apple.com/documentation/swift/string) object that is parsed to generate the returned [DateComponents](https://developer.apple.com/documentation/foundation/datecomponents) object.
     - returns: A  [DateComponents](https://developer.apple.com/documentation/foundation/datecomponents) object created by parsing `string`, or `nil` if string could not be parsed
      */
-    public func dateComponents(from string: String) -> DateComponents? {
-        var dateComponents: AnyObject? = nil
-        if getObjectValue(&dateComponents, for: string, errorDescription: nil) {
+    public func dateComponents(from string: String) throws -> DateComponents? {
+        var dateComponents: AnyObject?
+        var errorDescription: NSString?
+        if getObjectValue(&dateComponents, for: string, errorDescription: &errorDescription) {
             return dateComponents as? DateComponents
         }
+        if let errorDescription = errorDescription as? String {
+            if errorDescription == Error.fractionalValuesNotSupported.localizedDescription {
+                throw Error.fractionalValuesNotSupported
+            }
+            throw Error.description(message: errorDescription)
+        }
         
         return nil
     }
     
     public override func string(for obj: Any?) -> String? {
+        if let dateComponents = obj as? DateComponents {
+            return dateComponents.toISO8601Duration()
+        }
         return nil
     }
     
     public override func getObjectValue(_ obj: AutoreleasingUnsafeMutablePointer<AnyObject?>?, for string: String, errorDescription error: AutoreleasingUnsafeMutablePointer<NSString?>?) -> Bool {
```

Comment:
I think, with the new throwing, I would move the functionality to an extension on String itself. The formatter API would just become a proxy for this extension.